cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -118,26 +122,23 @@ public static KafkaCluster fromConfigMap(ConfigMap kafkaClusterCm) {
                 Labels.fromResource(kafkaClusterCm));
 
         Map<String, String> data = kafkaClusterCm.getData();
-        kafka.setReplicas(Integer.parseInt(data.getOrDefault(KEY_REPLICAS, String.valueOf(DEFAULT_REPLICAS))));
-        kafka.setImage(data.getOrDefault(KEY_IMAGE, DEFAULT_IMAGE));
-        kafka.setHealthCheckInitialDelay(Integer.parseInt(data.getOrDefault(KEY_HEALTHCHECK_DELAY, String.valueOf(DEFAULT_HEALTHCHECK_DELAY))));
-        kafka.setHealthCheckTimeout(Integer.parseInt(data.getOrDefault(KEY_HEALTHCHECK_TIMEOUT, String.valueOf(DEFAULT_HEALTHCHECK_TIMEOUT))));
+        kafka.setReplicas(getInteger(data, KEY_REPLICAS, DEFAULT_REPLICAS));
+        kafka.setImage(getString(data, KEY_IMAGE, DEFAULT_IMAGE));
+        kafka.setHealthCheckInitialDelay(getInteger(data, KEY_HEALTHCHECK_DELAY, DEFAULT_HEALTHCHECK_DELAY));
+        kafka.setHealthCheckTimeout(getInteger(data, KEY_HEALTHCHECK_TIMEOUT, DEFAULT_HEALTHCHECK_TIMEOUT));
 
-        kafka.setZookeeperConnect(data.getOrDefault(ENV_VAR_KAFKA_ZOOKEEPER_CONNECT, kafkaClusterCm.getMetadata().getName() + "-zookeeper:2181"));
+        kafka.setZookeeperConnect(kafkaClusterCm.getMetadata().getName() + "-zookeeper:2181");
 
         String metricsConfig = data.get(KEY_METRICS_CONFIG);
         kafka.setMetricsEnabled(metricsConfig != null);
         if (kafka.isMetricsEnabled()) {
             kafka.setMetricsConfig(new JsonObject(metricsConfig));
         }
 
-        String storageConfig = data.get(KEY_STORAGE);
-        kafka.setStorage(Storage.fromJson(new JsonObject(storageConfig)));
+        kafka.setStorage(getStorage(data, KEY_STORAGE));
 
-        String kafkaConfig = data.get(KEY_KAFKA_CONFIG);
-        if (kafkaConfig != null) {
-            kafka.setConfiguration(new KafkaConfiguration(new JsonObject(kafkaConfig)));
-        }
+        String kafkaConfig = getConfig(data, KEY_KAFKA_CONFIG);
+        kafka.setConfiguration(new KafkaConfiguration(new JsonObject(kafkaConfig)));
```

Comment:
Actually, the `Properties` class is kind of messy. I think it was not the best choice to use it in `AbstractConfiguration`. While it makes some things easier, it also brings a lot of limitations because it is opinionated and expects the use of the properties file only with Java. I'm afraid that one day we will need to get rid of ti again. We should not spread the use of Properties outside of `Configuration` classes into the `Model` classes.

Reply:
You're right that `Properties` shows its age and isn't fun to use. But currently the `AbstractConfiguration`'s subclasses are all Java things which take `.properties` files for their configuration: I don't think there's much point in trying to encapsulate that. When we have some component which consumes its config as YAML or XML or whatever else that would either be a different thing than an `AbstractConfiguration`, or we would at that point generify `AbstractConfiguration` to not care about the type.

I think it's perfectly reasonable to validate that the value in the CM conforms to something we can convert to a `.properties` file, and it's natural to do that by returning a `Properties` instance at the same time. The fact that `Properties` is old and ugly doesn't negate that.

Reply:
It is not about being old and ugly. The `Properties` class is designed primarily to write and load the properties from a single application. Not as a mean to create properties files. As a result it does things like comments, escaping and has a bad handling of default values.

AFAIK using `Properties` outside of the `Configuration` classes will anyway not give us any benefits compared to `Map<String, String>` which it wraps.

Reply:
> Not as a mean to create properties files.

I don't understand what you mean here. It *defines* the format for a `.properties` file.

> compared to `Map<String, String>` which it wraps.

I could live with that, I suppose.